gcloud/monitoring/query.py

```diff
@@ -0,0 +1,661 @@
+            just ``"resource.zone"`` will result in one time series per zone.
+            The default is to aggregate all of the time series into just one.
+
+        :rtype: :class:`Query`
+        :returns: The new query object.
+
+        .. _supported reducers:
+            https://cloud.google.com/monitoring/api/ref_v3/rest/v3/\
+            projects.timeSeries/list#Reducer
+        """
+        new_query = self.copy()
+        new_query._cross_series_reducer = cross_series_reducer
+        new_query._group_by_fields = group_by_fields
+        return new_query
+
+    def iter(self, headers_only=False, _page_size=None):
+        """Yield all time series objects selected by the query.
+
+        Note that the :class:`Query` object itself is an iterable, such that
+        the following are equivalent::
+
+            for timeseries in query: ...
+            for timeseries in query.iter(): ...
+
+        :type headers_only: boolean
+        :param headers_only:
+             Whether to omit the point data from the time series objects.
+
+        :type _page_size: integer or None
+        :param _page_size:
+            An optional positive number specifying the maximum number of
+            points to return per page. This can be used to control how far
+            the iterator reads ahead.
+
+        :rtype: iterator over :class:`~gcloud.monitoring.timeseries.TimeSeries`
+        :returns: Time series objects, containing points ordered from oldest
+            to newest.
+        :raises: :exc:`ValueError` if the query time interval has not been
+            specified.
+        """
+        if self._end_time is None:
+            raise ValueError('Query time interval not specified.')
+
+        path = '/projects/{project}/timeSeries/'.format(
+            project=self._client.project)
+
+        def _fragments():
+            page_token = None
+            while True:
+                params = self._build_query_params(
+                    headers_only=headers_only,
+                    page_size=_page_size,
+                    page_token=page_token,
+                )
+                response = self._client.connection.api_request(
+                    method='GET',
+                    path=path,
+                    query_params=params,
+                )
+                for info in response.get('timeSeries', []):
+                    yield TimeSeries._from_dict(info)
+
+                page_token = response.get('nextPageToken')
+                if not page_token:
+                    break
+
+        # The following use of groupby() relies on equality comparison
+        # of time series as (named) tuples.
+        for timeseries, fragments in itertools.groupby(
+                _fragments(),
+                lambda fragment: fragment.header()):
+            points = list(itertools.chain.from_iterable(
+                fragment.points for fragment in fragments))
+            points.reverse()  # Order from oldest to newest.
```

Comment:
It's a pity to have to use all these iterators (including in the `_fragments()` call) and then have to cast to a list just to reverse the points. Could the query be refined to avoid this? Are the number of points per fragment small-ish?